### PR TITLE
GRPO with JetStream offline engine

### PR DESCRIPTION
**Pull request by khatwanimohit**
# Description

This PR adds support for following things in GRPO
* Use JetStream offline engine as a sampler for inference
* Disaggregated training and inference using Pathways as backend 
* `end_to_end/tpu/test_grpo.sh` as a recipe for running `grpo_trainer` on v5p-256 for Llama3.3-70B
* Loading separate configs for training and inference 

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/123456
FIXES: #123456

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

**Review by richjames0 — APPROVED** [review]
I'd love to see more unit tests, e.g. around the changes to max_utils.py

File: MaxText/experimental/rl/README.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+This directory contains experimental code and documentation related to GRPO (Generalized Reinforcement Policy Optimization) within MaxText.
+
+## Getting Started
+
+#### Setup
+    
+Create your GKE cluster with Pathways https://cloud.google.com/ai-hypercomputer/docs/workloads/pathways-on-cloud/create-gke-cluster
+    
+
+## Run
```
Comment (A9isha) [inline MaxText/experimental/rl/README.md]: did you want to add something to Run?

File: requirements.txt [hunk 2]
```diff
@@ -39,5 +39,5 @@ transformers
 mlperf-logging@git+https://github.com/mlperf/logging.git
 google-jetstream@git+https://github.com/AI-Hypercomputer/JetStream.git
 jsonlines
-pathwaysutils==0.1.1
+pathwaysutils@git+https://github.com/AI-Hypercomputer/pathways-utils.git
```
Comment (A9isha) [inline requirements.txt:43]: we going from a stable release of pathwaysutils to running head (which might induce instabilities), is this required for the resharding API?

File: MaxText/experimental/rl/grpo_inference.yml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+inference_rollouts: 1
+
+decode_sampling_strategy: "weighted"
+decode_sampling_temperature: 0.9
+async_checkpointing: false
+
+return_log_prob: True
+
+add_bos: False
+add_eos: False
```
Comment (A9isha) [inline MaxText/experimental/rl/grpo_inference.yml]: nit: please add the newline in the end

File: MaxText/experimental/rl/grpo_trainer.py [hunk 4]
```diff
@@ -116,14 +140,17 @@ def grpo_loss_fn(model, config, data, dropout_rng, params, reference_params, is_
...
-         kl = exp(ref_logp - policy_logp) - (ref_logp - policy_logp) - 1.
-    3. Compute a scalar reward for each generated completion via reward_fn.
-    4. Group the rewards (each prompt yields “G = num_generations” completions), compute the mean and std,
+    2. Compute a scalar reward for each generated completion via reward_fn.
+    3. Group the rewards (each prompt yields “G = num_generations” completions), compute the mean and std,
        and then compute a normalized advantage.
-    5. Compute a per-token loss that is given by
-         - [exp(policy_logp - stop_gradient(policy_logp)) * advantage - beta * kl]
+    4. Compute a per-token loss that is given by
+         - [exp(policy_logp - old_logp) * advantage - beta * kl]
```
Comment (A9isha) [inline MaxText/experimental/rl/grpo_trainer.py:147]: this formula is for on-policy in particular

File: MaxText/experimental/rl/grpo_trainer.py [hunk 5]
```diff
@@ -116,14 +140,17 @@ def grpo_loss_fn(model, config, data, dropout_rng, params, reference_params, is_
...
+         where:
+         old_logp = stop_gradient(policy_logp) for on-policy training
        (the jax.lax.stop_gradient ensures that only the advantage contributes to gradients).
+         and actual old_logp obtained during sampling from the data["completions_logprobs"] for off-policy training.
```
Comment (A9isha) [inline MaxText/experimental/rl/grpo_trainer.py]: there are a few more coefficients for off-policy

File: MaxText/experimental/rl/grpo_trainer.py [hunk 6]
```diff
@@ -696,61 +528,157 @@ def train_loop(config, config_inference, recorder, state=None):
...
   p_train_step, p_eval_step = train_utils.jit_train_and_eval_step(
       config, model, mesh, state, state_mesh_shardings, train_step, eval_step, eval_data_iterator
   )
 
-  with mesh, nn_partitioning.axis_rules(config.logical_axis_rules):
-    shaped_batch = maxtext_utils.get_shaped_batch(config)
-    compiled = p_train_step.lower(state, shaped_batch, init_rng).compile()
-    compiled_stats = compiled.memory_analysis()
-    max_utils.print_compiled_memory_stats(compiled_stats)
+  # TODO: need to find a good way for get_shaped_batch
```
Comment (A9isha) [inline MaxText/experimental/rl/grpo_trainer.py]: would you want to remove these comments?

File: MaxText/experimental/rl/grpo_trainer.py [hunk 7]
```diff
@@ -696,61 +528,157 @@ def train_loop(config, config_inference, recorder, state=None):
...
+    worker_inference_engine,
+    worker_tokenizer_model,
+    worker_config_inference,
+    worker_config_train,
+    worker_data_buffer,
+    worker_data_buffer_lock,
+    worker_input_data_shardings,
+    engine_lock,
+  ):
+    with engine_lock:
```
Comment (A9isha) [inline MaxText/experimental/rl/grpo_trainer.py]: for my understanding - what issues arise if we don't use the lock?

File: MaxText/experimental/rl/grpo_utils.py [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+
+
+
+def dummy_reward_len(valid_seq_mask):
+  # adding a 1 because valid_seq_mask is actually one less than the number of valid tokens
+  reward = -abs(20 - (1 + jnp.sum(valid_seq_mask, axis=-1)))  # [BxG]
+  return reward
+
+
+def jaccard_reward_fn(tokens1, tokens2, vocab_size):
```
Comment (A9isha) [inline MaxText/experimental/rl/grpo_utils.py]: perhaps remove this since we are not using it

File: MaxText/experimental/rl/grpo_utils.py [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+  for i in range(num_groups):
+    current_group_dict = {}
+    start_index = i * global_batch_size_per_group
+    end_index = start_index + global_batch_size_per_group
+    for key, sliced_array in sliced_batch.items():
+      # Slice each group
+      current_group_dict[key] = sliced_array[start_index:end_index]
+    list_of_output_batches.append(current_group_dict)
+
+  return list_of_output_batches
```
Comment (A9isha) [inline MaxText/experimental/rl/grpo_utils.py]: nit: please add the new line

File: MaxText/tests/grpo_trainer_correctness_test.py [hunk 10]
```diff
@@ -114,6 +117,7 @@ def setUp(self):
         run_name="unit_test_grpo_trainer",
         tokenizer_path=os.path.join(os.path.dirname(PKG_DIR), "assets", "llama3.1-tokenizer"),
         enable_checkpointing=False,
+        train_data_columns='prompt',
```
Comment (A9isha) [inline MaxText/tests/grpo_trainer_correctness_test.py]: where are we reading the `completion` column from TLDR dataset? I know for the current reward function we don't need it, but I am thinking it should be read somewhere as an example

File: MaxText/experimental/rl/README.md (hunk 1 above)
Comment (A9isha) [reply MaxText/experimental/rl/README.md]: perhaps end_to_end/tpu/test_grpo.sh

File: end_to_end/tpu/test_grpo.sh [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+per_device_batch_size=${TRAINING_PER_DEVICE_BATCH_SIZE} num_generations=${NUM_GENERATIONS} steps=${STEPS}"
+
+INFERENCE_ARGS="run_name=grpo scan_layers=false \
+per_device_batch_size=${INFERENCE_PER_DEVICE_BATCH_SIZE} \
+ici_data_parallelism=${NUM_SAMPLERS} ici_tensor_parallelism=${DEVICES_PER_SAMPLER}"
+
+JAX_PLATFORMS=proxy JAX_BACKEND_TARGET=grpc://127.0.0.1:29000 ENABLE_PATHWAYS_PERSISTENCE='1' \
+    python3 -m MaxText.experimental.rl.grpo_trainer MaxText/experimental/rl/grpo.yml  \
+    ${COMMON_ARGS} ${TRAINING_ARGS} MaxText/experimental/rl/grpo_inference.yml \
+    ${COMMON_ARGS} ${INFERENCE_ARGS}
```
Comment (A9isha) [inline end_to_end/tpu/test_grpo.sh]: nit: please add the newline

File: MaxText/tests/grpo_trainer_correctness_test.py [hunk 12]
```diff
@@ -203,16 +215,15 @@ def test_grpo_trainer_correctness(self):
...
-        donate_argnums=(0,),
-    )
-    # pylint: disable=not-callable
-    engine_data = p_generate_completions(engine_data, {"params": state.params["params"]}, rng)
+    
+    input_data = []
+    for i, d in enumerate(engine_data[self.config.train_data_columns]):
+      input_data.append(InputData(id=f"input_{i}", tokens=np.array(d), true_length=np.array(data[f"{self.config.train_data_columns}_true_length"][i])[0]))
+    
+    results = self.inference_engine.batch_inference(input_data)
```
Comment (A9isha) [inline MaxText/tests/grpo_trainer_correctness_test.py:228]: for completeness(I know we have disabled it from running), we should also make these changes to MaxText/tests/integration_tests/grpo_correctness.py

File: MaxText/tests/grpo_trainer_correctness_test.py (hunk 10 above)
Comment (khatwanimohit) [reply MaxText/tests/grpo_trainer_correctness_test.py]: All the columns are read as part of the input pipeline and are part of the example_batch when loaded.

File: MaxText/experimental/rl/grpo_trainer.py (hunk 7 above)
Comment (khatwanimohit) [reply MaxText/experimental/rl/grpo_trainer.py]: We are running generate_completions function in a thread, so introducing a lock before we perform operations makes them thread safe

File: MaxText/experimental/rl/grpo_trainer.py (hunk 5 above)
Comment (khatwanimohit) [reply MaxText/experimental/rl/grpo_trainer.py]: I have improved the comment, PTAL

File: requirements.txt (hunk 2 above)
Comment (khatwanimohit) [reply requirements.txt:43]: Yes this is required for resharding API, I will revert back to stable once it is released

File: MaxText/experimental/rl/README.md (hunk 1 above)
Comment (khatwanimohit) [reply MaxText/experimental/rl/README.md]: Done

File: MaxText/tests/grpo_trainer_correctness_test.py (hunk 10 above)
Comment (A9isha) [reply MaxText/tests/grpo_trainer_correctness_test.py]: where do we provide the column names for the dataset?

**Review by A9isha — APPROVED** [review]
Thank you Mohit!

File: MaxText/tests/grpo_trainer_correctness_test.py (hunk 10 above)
Comment (khatwanimohit) [reply MaxText/tests/grpo_trainer_correctness_test.py]: All the columns are read so we don't need to provide any column names while reading the dataset

File: requirements.txt [hunk 13]
```diff
@@ -40,7 +40,7 @@ tiktoken>=0.9.0
 transformers>=4.52.4
 transformer-engine[jax]>=2.1.0
 jsonlines>=4.0.0
-pathwaysutils>=0.1.1
+pathwaysutils@git+https://github.com/AI-Hypercomputer/pathways-utils.git
```
Comment (bvandermoon) [inline requirements.txt:43]: Is this also needed in these files?

* `constraints_gpu.txt`
* `requirements_with_jax_ai_image.txt`
* `requirements_with_jax_stable_stack_0_6_1_pipreqs.txt`
